docs/src/docs/asciidoc/user/resilience.adoc

```diff
@@ -0,0 +1,189 @@
+---
+---
+= Resilience
+
+*Note:* This is not the current state of things. It is the current ongoing effort to improve cluster resilience.
+
+== What can go wrong
+
+* Server fails to get an answer erratically because of network or something
+* Server fails to send messages to the client erratically
+* Server goes down
+* Server failover
+* Loader-writer backend fails
+
+== Strong vs Eventual
+
+Strong means: If the client ask for a value that was updated by some other server, it will always get the
+latest value.
+
+In particular, it means that as soon as an update from the server might be missed, we need to clear
+the caching tier. It also means that if we fail to set a value to the authoritative tier, we can't use
+the cache anymore. On this key at least.
+
+Eventual means the correct value will eventually be there. So if we loose contact with the server, it is possible to keep
+using a possibly stale cache entry. This means that an client with eventual consistency shouldn't need to clear it's cache
+at all. However, it means two things:
+
+* We need to be able to update the caching tier in that case
+* We need to get updates from the server when the situation resumes
+
+=== Being too eventual
+
+We can argue that if a client stays disconnected for too long, it means the caching tier content is now too obsolete. There
+are two ways to handle that
+
+* Expiration should take care of it. So entries should have an expiration set to their expected time validity.
+* We configure "too long" which will clear the cache
+
+== Clustering
+
+When using a distributed cache, your Ehcache client will connect to a remote Terracotta server (or multiple in case
+of stripping) which act as the authoritative tier.
+
+This server data can be replicated to other servers called mirrors. How your client will handle being cut from
+the cluster is defined in `ClusteringServiceConfiguration`.
+
+=== Timeouts
+
+If a client doesn't receive an answer in the configured time, it will timeout. This can happen for different reasons
+including a network cut, a long full GC or a server down.
+
+Three different types of timeouts have been defined.
+
+* Read: Operations reading data from the server
+* Mutative: Operations modifying data on the server
+* Lifecycle: Operations related to the lifecycle like creating and destroying a cache or simply connecting to the server
+
+By default, all are set to 5 seconds.
+
+=== Retries
+
+When a client timeouts, it might retry right away. That are two parameters for that:
+
+* Number of retries (default is 0)
+* Sleep time between retries (default is 0)
+
+Note that for each retry, the timeout will be enforced. So, far example, 3 retries with a timeout of 5 seconds and a sleep
+of 1 second will mean 20 seconds before returning.
+
+*Note to myself:* Should the retry be on the resilience strategy?
```

Comment:
I still don't understand why the user should care about retries?  If the user sets a read timeout of 5 seconds he would expect us to do whatever is prudent to honor the read within that time.  If that means retries then that makes sense, but that would then be an implementation concern.  How would someone decide between:

* 20 second timeout, no retry.
* 5 second timeout, 3 retries.

To me which of those is right is really only known to the implementation, and even then it's only really known when faced with the failure since different failures might be best tackled with different retry approaches.

Reply:
Ok. Agreed. We can expose it later anyway if it becomes useful for any reason.